### DISCUSSION TITLE
x-if vs x-show

### DISCUSSION THREAD

**Discussion by usernotnull** [alpinejs/alpine#1086] [1. Help] [answered]
Between x-if and x-show, is there a difference in performance? Manipulating the display none vs doc fragments...
I'm finding myself always using x-show because it involves less typing and wondering what's the use of the x-if.

**Comment by ryangjchandler** [comment, marked as answer]
@RJFares - In general, modifying an element's style list will be more performant than importing document nodes and initialising them with Alpine, especially if the content inside of `x-if` is quite deep (multiple nested elements), since Alpine will need to walk this each time the element is inserted in the DOM.

Unless there's a particular reason why you wouldn't want the element being conditionally displayed visible in the DOM at all times, I'd always go with `x-show` (personally).

**Reply by usernotnull** [reply]
So same as docs mention, which is if for whatever reason x-show didn't work, try x-if... Adding that there's a performance cost.
Thanks

**Reply by ryangjchandler** [reply]
> So same as docs mention, which is if for whatever reason x-show didn't work, try x-if... Adding that there's a performance cost.
> Thanks

If you think there's a problem with `x-show`, feel free to open an issue. If you're just looking for some help, the [Discord](https://discord.com/invite/CGmj5nq) is also a great place to be.

**Reply by HugoDF** [reply]
> So same as docs mention, which is if for whatever reason x-show didn't work, try x-if... Adding that there's a performance cost.
> Thanks

One reason you would go for `x-if` is if you're doing deep object access in your template eg. `obj.a.b` but sometime `obj.a` is undefined/null.

With `x-show="obj.a"` it would still evaluate the directives inside the element so your might get an error like. "cannot read property of undefined".

With `x-if="obj.a"` you can guard against the object being null/undefined.

**Comment by dfallman** [comment]
This is interesting. I experience quite a performance difference when using `x-show` over `x-if` in a long list of lines (1k+) generated by `x-for` where every line has ~10+ options for things to show and hide. I previously had these as `x-if`s but changing them to `x-show` seems to speed up everything quite a bit actually. 

I have to admit I completely missed this in the docs (is it still there?)

Another related question: is there a good/reliable way to performance test x-for loops, i.e. time how long it takes to redraw the list?

**Reply by ekwoka** [reply]
Well, Alpine is included in framework benchmarks, which includes large table changes. So that is a decently reliable source (and that repo has all you need to run the benchmarks yourself).

When this thread was opened, Alpine was on V2, and V3 has some major under the hood changes. I'm not sure how much they impact these differences for x-if and x-show, but yes, generally the act of toggling will be faster with x-show. x-if is beneficial when the toggles themselves are less frequent, like nearly static, and have major logic in their directives or other aforementioned "type-guarding".

x-show may take more time during initial evaluations, but it means it won't need to do it later, or again.

Often, if any of them is noticeably having issues, there is likely something else going on
